### PR TITLE
fix the comment for self are swallowed

### PR DESCRIPTION
**Pull request by rchaser53**
related: https://github.com/rust-lang/rustfmt/issues/3198

This is a bug fix, so I don't need to add Version Gate. Did I get that right?

### CODE REVIEW

File: src/items.rs [hunk 1]
```diff
@@ -2289,11 +2294,26 @@ fn rewrite_args(
...
-    // on rust issue #27522.
+    let mut pre_comment_str = "";
+    let mut post_comment_str = "";
     let min_args = explicit_self
         .and_then(|explicit_self| rewrite_explicit_self(explicit_self, args, context))
         .map_or(1, |self_str| {
+            pre_comment_str = context
+                .snippet_provider
+                .span_to_snippet(mk_sp(span.lo(), args[0].pat.span.lo()))
+                .unwrap();
```
Comment (topecongiro) [inline src/items.rs]: Same as https://github.com/rust-lang/rustfmt/pull/3415#discussion_r259622767.

**Comment by rchaser53** [comment]
Thank you for the review. I fixed it.

File: src/items.rs [hunk 2]
```diff
@@ -2357,20 +2371,38 @@ fn rewrite_args(
...
         arg_items.extend(more_items);
     }
 
+    let arg_items_len = arg_items.len();
     let fits_in_one_line = !generics_str_contains_newline
         && (arg_items.is_empty()
-            || arg_items.len() == 1 && arg_item_strs[0].len() <= one_line_budget);
+            || arg_items_len == 1 && arg_item_strs[0].len() <= one_line_budget);
+
+    let has_second_pre_comment = arg_items_len > 1 && arg_items[1].pre_comment.is_some();
```
Comment (scampi) [inline src/items.rs]: Can you describe what this does and why it's needed ?

Comment (rchaser53) [reply src/items.rs]: Please see this [link](https://travis-ci.com/rust-lang/rustfmt/jobs/180096057)

Second arg pre comment also includes first arg post comment in linux environment.
So needs to remove second pre comment if first arg post comment exists.

Comment (rchaser53) [reply src/items.rs]: I added the comment to source code.

**Comment by rchaser53** [comment]
I think I should fix other place, so add wip to title.

**Comment by rchaser53** [comment]
I improved the implements

File: src/items.rs [hunk 3]
```diff
@@ -2310,14 +2324,18 @@ fn rewrite_args(
...
     if args.len() >= min_args || variadic {
         let comment_span_start = if min_args == 2 {
-            let second_arg_start = if arg_has_pattern(&args[1]) {
-                args[1].pat.span.lo()
+            let remove_comma_byte_pos = context
+                .snippet_provider
+                .span_after(mk_sp(args[0].ty.span.hi(), args[1].pat.span.lo()), ",");
+            let first_post_and_second_pre_span =
+                mk_sp(remove_comma_byte_pos, args[1].pat.span.lo());
+            if count_newlines(context.snippet(first_post_and_second_pre_span)) > 0 {
```
Comment (scampi) [inline src/items.rs]: Can you add a test where the comment spans multiple lines so that this conditional is checked ?

File: src/items.rs [hunk 4]
```diff
@@ -2357,18 +2375,30 @@ fn rewrite_args(
...
 
+    let arg_items_len = arg_items.len();
     let fits_in_one_line = !generics_str_contains_newline
         && (arg_items.is_empty()
-            || arg_items.len() == 1 && arg_item_strs[0].len() <= one_line_budget);
+            || arg_items_len == 1 && arg_item_strs[0].len() <= one_line_budget);
 
-    for (item, arg) in arg_items.iter_mut().zip(arg_item_strs) {
+    for (index, (item, arg)) in arg_items.iter_mut().zip(arg_item_strs).enumerate() {
+        if index == 0 && explicit_self.is_some() {
```
Comment (scampi) [inline src/items.rs]: Can you add a comment in the code describing what this block does ?

**Comment by rchaser53** [comment]
Thank you fot the review.
I added the comment and some tests.

**Comment by scampi** [comment]
Looking good! @topecongiro are you OK with the changes ?

**Comment by scampi** [comment]
build failure is unrelated, it's about clippy failing to compile (before running rustfmt).

**Comment by topecongiro** [comment]
LGTM, thanks!

**Pull request merged**